internal/cache/file/cache_handler.go

```diff
@@ -0,0 +1,224 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/file/downloader"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/util"
+	"github.com/googlecloudplatform/gcsfuse/internal/locker"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+)
+
+// CacheHandler Responsible for managing fileInfoCache as well as fileDownloadManager.
+type CacheHandler struct {
+	// fileInfoCache contains the reference of fileInfo cache.
+	fileInfoCache *lru.Cache
+
+	// jobManager contains reference to a singleton jobManager.
+	jobManager *downloader.JobManager
+
+	// Directory location which would contain local file to keep the downloaded data.
+	cacheLocation string
+
+	// filePerm parameter specifies the permission behaviour to determine which users
+	// will have access to the cache file.
+	filePerm os.FileMode
+
+	// Guards the handling of cache entry insertion while creating the CacheHandle.
+	// Also guards post eviction logic E.g. cancelling and deletion of async download job,
+	// deletion of local file containing downloaded data.
+	mu locker.Locker
+}
+
+func NewCacheHandler(fileInfoCache *lru.Cache, jobManager *downloader.JobManager, cacheLocation string, filePerm os.FileMode) *CacheHandler {
+	return &CacheHandler{
+		fileInfoCache: fileInfoCache,
+		jobManager:    jobManager,
+		cacheLocation: cacheLocation,
+		filePerm:      filePerm,
+		mu:            locker.New("FileCacheHandler", func() {}),
+	}
+}
+
+func (chr *CacheHandler) createLocalFileReadHandle(objectName string, bucketName string) (*os.File, error) {
+	fileSpec := data.FileSpec{
+		Path: util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(bucketName, objectName)),
+		Perm: chr.filePerm,
+	}
+
+	return util.CreateFile(fileSpec, os.O_RDONLY)
+}
+
+func (chr *CacheHandler) cleanUpEvictedFile(fileInfo *data.FileInfo) error {
+	key := fileInfo.Key
+	_, err := key.Key()
+	if err != nil {
+		return fmt.Errorf("error while performing post eviction: %v", err)
+	}
+
+	// Removing Job doesn't delete the job object itself but invalidates the job and
+	// hence it's possible that some existing cache handle will have reference to older
+	// job object, but that job object will have INVALID status.
+	chr.jobManager.RemoveJob(key.ObjectName, key.BucketName)
+
+	localFilePath := util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(key.BucketName, key.ObjectName))
+	err = os.Remove(localFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logger.Warnf("while clean up, filePath should exist: %v", err)
```

Comment:
cleanUpEvictedFile: file was not present at the time of clean up.

Let's start the error string with function name at all places.